pkg/algorithm/sku_scheduler.go

```diff
@@ -0,0 +1,517 @@
+					bestAffinity,
+					bestAffinityCells,
+					bestAffinityCellIndices,
+					optimalAffinity)
+				if foundOptimalAffinity {
+					// early stop: return if the solution is optimal (i.e., all buddies)
+					return bestAffinityCells
+				}
+			} else {
+				searchCellIndex++
+			}
+			freeCellIndex++
+		}
+		searchCellIndex--
+		if searchCellIndex < 0 {
+			if bestAffinity == highestLevel {
+				// Unreachable
+				panic(fmt.Sprintf("Assert Failure: failed to allocate %v cells in cell %v", pod.CellsPerPod.CellNumber, within.address))
+			}
+			return bestAffinityCells
+		}
+		freeCellIndex = currentCellIndices[searchCellIndex] + 1
+	}
+}
+
+func (s *skuScheduler) getNodeLevel() CellLevel {
+	for l := CellLevel(1); l <= CellLevel(len(s.chainCellList)); l++ {
+		if s.chainCellList[l][0].AtOrHigherThanNode() {
+			return l
+		}
+	}
+	return -1
+}
+
+// getFreeCellsAtLevel collects free cells and preemptible cells at given level according to the priority.
+func getFreeCellsAtLevel(
+	cell Cell,
+	level CellLevel,
+	priority CellPriority,
+	allocatedCells CellList,
+	freeCells CellList,
+	preemptibleCells CellList) (
+	CellList, CellList) {
+
+	if cell.GetLevel() > level {
+		for _, c := range cell.GetChildren() {
+			freeCells, preemptibleCells = getFreeCellsAtLevel(
+				c, level, priority, allocatedCells, freeCells, preemptibleCells)
+		}
+	} else if cell.GetLevel() == level {
+		isAllocated := false
+		for _, c := range allocatedCells {
+			if isAncestor(cell, c) || isAncestor(c, cell) {
+				isAllocated = true
+				break
+			}
+		}
+		if !isAllocated {
+			if cell.GetPriority() == freePriority {
+				freeCells = append(freeCells, cell)
+			} else if cell.GetPriority() < priority {
+				preemptibleCells = append(preemptibleCells, cell)
+			}
+		}
+	}
+	return freeCells, preemptibleCells
+}
+
+// checkOptimalAffinityForCells checks if the currently picked cells have the lowest LCA.
+// It also checks if the solution is optimal (if the leaf cells are all buddies).
+func checkOptimalAffinityForCells(
+	affinity CellLevel,
+	freeCells CellList,
+	currentCellIndices []int32,
+	bestAffinity CellLevel,
+	bestAffinityCells CellList,
+	bestAffinityCellIndices []int32,
+	optimalAffinity CellLevel) (CellLevel, bool) {
+
+	if affinity < bestAffinity {
+		copy(bestAffinityCellIndices, currentCellIndices)
+		for i := 0; i < len(currentCellIndices); i++ {
+			bestAffinityCells[i] = freeCells[currentCellIndices[i]]
+		}
+		if affinity == optimalAffinity {
+			return affinity, true
+		} else {
+			return affinity, false
+		}
+	}
+	return bestAffinity, false
+}
+
+func (s *skuScheduler) createSkuClusterView(
+	within *skuCell,
+	withinLevel CellLevel,
+	priority CellPriority) skuClusterView {
+
+	cv := skuClusterView{}
+	for l := withinLevel; l >= CellLevel(1); l-- {
+		for _, c := range s.chainCellList[l] {
+			if (within != nil && !isAncestor(within.cell, c)) ||
+				cv.contains(ancestorNoHigherThanLevel(withinLevel, c)) {
+				continue
```

Comment:
Will "cv.contains() continue" cause some cells with more free gpus are excluded in the final cv?